---
title: "Review Project Key"
author: Wade Boohar
date: 11/12/24
updated: 11/12/24
---

```{r setup}
    knitr::opts_knit$set(root.dir = normalizePath("~/490_cluster/analysis_data")) 
```

```{r}
library(BiocManager)
library(TCGAbiolinks)
library(SummarizedExperiment)
library(DESeq2)
library(survival)
library(survminer)
library(ggplot2)
library(EnhancedVolcano)

if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")
BiocManager::install("sesame")
BiocManager::install("sesameData")
BiocManager::install("limma")
library(sesame)
library(sesameData)
library(limma)
```



### LOAD IN DATA FOR ALL CLINICAL AND OMICS

# clinical - we are only using rad and drugs, we will be using rna_se@colData for our clinical data b/c it has metastasis status which is our main variable of interest
```{r}
clin_query <- GDCquery(project = "TCGA-SKCM",
                       data.category = "Clinical",
                       data.type = "Clinical Supplement",
                       data.format = 'BCR Biotab')
GDCdownload(clin_query)
clinical.BCRtab.all <- GDCprepare(clin_query)
rad <- clinical.BCRtab.all$clinical_radiation_skcm[-c(1,2),]
drug <- clinical.BCRtab.all$clinical_drug_skcm[-c(1,2),]
```

# rna_se
```{r}
rna_query <- GDCquery(project ="TCGA-SKCM",
                      data.category = "Transcriptome Profiling",
                      data.type = "Gene Expression Quantification",
                      workflow.type = "STAR - Counts")

GDCdownload(rna_query)
```
```{r}
rna_se <- GDCprepare(rna_query)
```


# create our dataframes from rna_se object and name appropriately
```{r}
rna_clinical <- rna_se@colData
rna_clinical <- data.frame(rna_clinical)

rna_genes <- rna_se@rowRanges@elementMetadata
rna_genes <- data.frame(rna_genes)

rna_counts <- rna_se@assays@data$unstranded
rna_counts <- data.frame(rna_counts)

rownames(rna_counts) <- rna_genes$gene_id
colnames(rna_counts) <- rna_clinical$patient
```

define metastasis status, remove patients missing data
```{r}
rna_clinical$definition <- factor(rna_clinical$definition, levels=c('Metastatic', 'Primary solid Tumor'))
metastasis_na_mask <- !is.na(rna_clinical$definition)
rna_clinical <- rna_clinical[metastasis_na_mask,]
rna_counts <- rna_counts[, metastasis_na_mask]
rna_clinical$metastasis <- ifelse(rna_clinical$definition == 'Metastatic', 'Metastatic', 'Non_Metastatic')
```


```{r}
query <- GDCquery(project = "TCGA-SKCM",
                  data.category = "DNA Methylation",
                  data.type = "Methylation Beta Value",
                  platform = "Illumina Human Methylation 450",
                  data.format = "Dataframe")
```
```{r}
GDCdownload(query)
```

```{r}
methylation450 <- GDCprepare(query)
```


```{r}

vital_status_mask <- ifelse(rna_clinical$vital_status %in% c('Alive', 'Dead'), T, F)
rna_clinical_km <- rna_clinical[vital_status_mask ,]
rna_clinical_km$death_event = ifelse(rna_clinical_km$vital_status=='Dead', T, F)

rna_clinical_km$days_to_death = as.numeric(rna_clinical_km$days_to_death)

rna_clinical_km$days_to_last_follow_up = as.numeric(rna_clinical_km$days_to_last_follow_up)

days_death_mask <- !is.na(rna_clinical_km$days_to_death)
rna_clinical_km$survival_time[days_death_mask] <- rna_clinical_km[days_death_mask, 'days_to_death']

follow_up_mask <- !is.na(rna_clinical_km$days_to_last_follow_up)
rna_clinical_km$survival_time[follow_up_mask] <- rna_clinical_km[follow_up_mask, 'days_to_last_follow_up']

survival_time_mask <- !is.na(rna_clinical_km$survival_time)
rna_clinical_km <- rna_clinical_km[survival_time_mask, ]
```

```{r}
rna_clinical_km$metastasis <- factor(rna_clinical_km$metastasis, levels=c('Metastatic', 'Non_Metastatic'))

survival_object <- Surv(time = rna_clinical_km$survival_time,
                        event = rna_clinical_km$death_event)

fit_object <- survfit(survival_object ~ metastasis, data = rna_clinical_km)

survplot <- ggsurvplot(fit_object,
                       pval=TRUE,
                       ggtheme = theme(plot.margin = unit(c(1,1,1,1), "cm")),
                       legend = 'right')

KM_plot <- survplot$plot + theme_bw() + theme(axis.title = element_text(size=20), 
                                              axis.text = element_text(size=16),
                                              legend.title = element_text(size=14),
                                              legend.text = element_text(size=12))

KM_plot
```

```{r}
# create column rna_clinical$radiation where if the patient had radiation treatment, T
rna_clinical$radiation <- ifelse(rna_clinical$patient %in% rad$bcr_patient_barcode, 1, 0)

# create column rna_clinical$chemotherapy where if the patient had chemotherapy, T
unique(drug$pharmaceutical_therapy_type)
all_chemo <- drug[drug$pharmaceutical_therapy_type == 'Chemotherapy',]
rna_clinical$chemotherapy <- ifelse(rna_clinical$patient %in% all_chemo$bcr_patient_barcode, 1, 0)

# create column rna_clinical$immunotherapy
all_immuno <- drug[drug$pharmaceutical_therapy_type == 'Immunotherapy',]
rna_clinical$immunotherapy <- ifelse(rna_clinical$patient %in% all_immuno$bcr_patient_barcode, 1, 0)

# create column rna_clinical$vaccine
all_vaccine <- drug[drug$pharmaceutical_therapy_type == 'Vaccine',]
rna_clinical$vaccine <- ifelse(rna_clinical$patient %in% all_vaccine$bcr_patient_barcode, 1, 0)


# create column rna_clinical$molecular_therapy
all_mol <- drug[drug$pharmaceutical_therapy_type == 'Targeted Molecular therapy',]
rna_clinical$molecular_therapy <- ifelse(rna_clinical$patient %in% all_mol$bcr_patient_barcode, 1, 0)

race_mask <- ifelse(rna_clinical$race %in% c('asian', 'black or african american', 'white'), T, F)
rna_clinical <- rna_clinical[race_mask, ]
rna_counts <- rna_counts[, race_mask]

vital_status_mask <- ifelse(rna_clinical$vital_status %in% c('Alive', 'Dead'), T, F)
rna_clinical <- rna_clinical[vital_status_mask, ]
rna_counts <- rna_counts[, vital_status_mask]

unique(rna_clinical$metastasis)
unique(rna_clinical$race)
unique(rna_clinical$vital_status)
unique(rna_clinical$gender)
```

# turn all necessary columns into factors
```{r}
rna_clinical$metastasis <- factor(rna_clinical$metastasis)
rna_clinical$gender <- factor(rna_clinical$gender)
rna_clinical$vital_status <- factor(rna_clinical$vital_status)
rna_clinical$molecular_therapy <- factor(rna_clinical$molecular_therapy)
rna_clinical$immunotherapy <- factor(rna_clinical$immunotherapy)
rna_clinical$chemotherapy <- factor(rna_clinical$chemotherapy)
rna_clinical$vaccine <- factor(rna_clinical$vaccine)
rna_clinical$radiation <- factor(rna_clinical$radiation)
rna_clinical$race <- factor(rna_clinical$race)
```

```{r}
# use rowSums() to create a list with the total number of counts of each gene
row_sums <- rowSums(rna_counts)
  
#create a boolean mask where genes with < 20 total counts are FALSE, and genes with >= 20 total counts are TRUE
low_counts_mask <- ifelse(row_sums < 20, F, T)
  
# rewrite the rna_counts df, subsetting for only genes with >= 10 total counts
rna_counts <- rna_counts[low_counts_mask,]
  
#update rna_genes with the low_counts_mask
rna_genes <- rna_genes[low_counts_mask,]

colnames(rna_counts) <- rna_clinical$barcode
```

```{r}
dds <- DESeqDataSetFromMatrix(countData = rna_counts,
                                colData = rna_clinical,
                                design = ~gender + vital_status + molecular_therapy + immunotherapy + chemotherapy + vaccine + radiation + race + metastasis)

dds_obj <- DESeq(dds) # note: this will likely take a long time (ie 45 minutes to 2 hours)

resultsNames(dds_obj)  # see what comparisons got run

results <- results(dds_obj, format = "DataFrame", contrast = c("metastasis", "Metastatic", "Non_Metastatic")) # this is case sensitive so be careful to match it with your age_category factors closely!

#We're not interested in all of these columns, so use data.frame() to write results with the following columns in this order: gene name (from rna_genes), gene id (from results rownames), log2foldchange, pvalue, padj, -log10(padj).
#Rename the columns using colnames(). Note: to get -log10, simply write "-log10(fill in here)".

results <- data.frame(rna_genes$gene_name, rownames(results), results$log2FoldChange, results$pvalue, results$padj, -log10(results$padj))

colnames(results) <- c("gene_name", "gene_id", "log2FC", "pval", "padj", "-log10(padj)") ## FIX column names

# add gene_id as the row names for results
row.names(results) <- results$gene_id

#select rows (genes) that have a padj value < 0.05.
sig_results <- as.data.frame(results[results$padj < 0.05,])

#Now, sort the data frame by log2foldchange, descending. To do this, use the order() function. The order() function returns a vector of the sorted indexes of your chosen data, from smallest to largest. For example:

#Use the logic above as well as sig_results to create a dataframe called up_reg_results. Remember that upregulated genes will have a statistically significant fold change greater than a positive number threshold (eg. log2FoldChange > 1 and adjusted p-values < 0.05). 

#This df should be sorted by log2foldchange descending (take a look at the decreasing argument). Note that this can also be done in one line by bypassing the use of order_indexes as an intermediate variable.

up_reg_results <- sig_results[order(sig_results$log2FC, decreasing = TRUE),]

# Up_reg_results should only contain the genes where the log2foldchange is > 1 (ie genes that are 2x or greater expressed in one population over the other).

up_reg_results <- up_reg_results[up_reg_results$log2FC > 1,]
  
#Repeat the same steps to create a sorted data frame called down_reg_results (this time select for all genes where log2foldchange is < -1 and sort it ascending by log2foldchange).

down_reg_results <- sig_results[order(sig_results$log2FC),]
down_reg_results <- down_reg_results[down_reg_results$log2FC < -1,]

EnhancedVolcano(results,
                lab = results$gene_name,
                x = 'log2FC',
                y = 'pval')


```


```{r}
methylation_clinical <- as.data.frame(methylation450@colData)
betas <- as.data.frame(methylation450@assays@data@listData)
cpg_sites <- as.data.frame(methylation450@rowRanges@elementMetadata)

column_mask <- ifelse(colnames(methylation_clinical) %in% c('treatments', 'primary_site', 'disease_type'), F, T)
methylation_clinical <- methylation_clinical[,column_mask]
write.csv(methylation_clinical, 'sckm_methylation_clinical.csv')


site_mask <- !grepl('-', cpg_sites$gene) & !grepl(';', cpg_sites$gene) & !is.na(cpg_sites$gene) & complete.cases(betas)
betas <- betas[site_mask,]
cpg_sites <- cpg_sites[site_mask,]

#write.csv(betas, 'skcm_methylation_betas.csv')
#write.csv(cpg_sites, 'skcm_cpg_sites.csv')
```


```{r}
#methylation_clinical <- read.csv('skcm_methylation_clinical.csv')
#betas <- read.csv('skcm_methylation_betas.csv')
#cpg_sites <- read.csv('skcm_cpg_sites.csv')
```

```{r}
other_vals_mask <- methylation_clinical$definition %in% c('Additional Metastatic', 'Solid Tissue Normal')
methylation_clinical <- methylation_clinical[!other_vals_mask,]
betas <- betas[, !other_vals_mask]

methylation_clinical$metastatic <- methylation_clinical$definition == 'Metastatic'

mval <- t(apply(betas, 1, function(x) log2(x/(1-x))))

design <- model.matrix(~ metastatic, data = methylation_clinical)
fit <- lmFit(mval, design)
fit2 <- eBayes(fit)
```

```{r}

dat <- data.frame(foldchange = fit[["coefficients"]][,2], logPvalue =  -log10(p.adjust(fit2[["p.value"]][,2],method='BY')), geneName = cpg_sites$gene)
dat$threshold <- as.factor(abs(dat$foldchange) < 1)

#Visualization
cols <- c("TRUE" = "grey", "FALSE" = "blue")
ggplot(data=dat, aes(x=foldchange, y = logPvalue, color=threshold)) +
  geom_point(alpha=.2, size=0.6) +
  scale_colour_manual(values = cols) +
  geom_vline(xintercept = 1, colour="#990000", linetype="dashed") + 
  geom_vline(xintercept = - 1, colour="#990000", linetype="dashed") +
  geom_hline(yintercept = 2, colour = "#990000", linetype="dashed") +
  theme(legend.position="none") +
  xlab("Fold Change") +
  ylab("-log10 p value") +
  theme_bw() +
  theme(legend.position = "none")
```


```{r}
downregulated <- results[(results$log2FC < -1), 'gene_name']
hypermethylated <- dat[dat$foldchange > 0.75, 'geneName']
interest_genes <- intersect(downregulated, hypermethylated)
```




```{r}
GENE<-"HAS3"

gene_counts_mask <- rna_genes$gene_name == GENE
gene_betas_mask <- cpg_sites$gene == GENE

rna_clinical_tumor <- rna_clinical$definition == "Primary solid Tumor"
methylation_clinical_tumor <- methylation_clinical$definition == "Primary solid Tumor"

rna_clinical_metastatic <- rna_clinical$definition == "Metastatic"
methylation_clinical_metastatic <- methylation_clinical$definition == "Metastatic"

rna_tumor <- as.numeric(rna_counts[gene_counts_mask, rna_clinical_tumor])
methylation_tumor <- (betas[gene_betas_mask, methylation_clinical_tumor])

rna_metastatic <- as.numeric(rna_counts[gene_counts_mask, rna_clinical_metastatic])
methylation_metastatic <- (betas[gene_betas_mask, methylation_clinical_metastatic])
```

```{r}
boxplot(rna_metastatic, rna_tumor, main=GENE, xlab='Group', ylab='Counts', names=c('Metastatic', 'Non_Metastatic'))
```
```{r}
plot(1:dim(methylation_metastatic)[1], apply(methylation_tumor, 1, mean), col='blue', type='h', lwd=3, ylab='Average Beta', xlab='CpG sites', main=GENE)
lines(1.3:(dim(methylation_metastatic)[1]+0.3), apply(methylation_metastatic, 1, mean), col='red', type='h', lwd=3)

# blue is non-metastatic, red is metastatic
```







